core/services/vrf/v2/listener_v2_log_listener_test.go

```diff
@@ -0,0 +1,1021 @@
+		require.NoError(t, err1)
+		th.Client.Commit()
+	}
+
+	// Blocks till now: 2 (in SetupTH) + 2 (empty blocks) + 1 (VRF req block) + 5 (EmitLog blocks) = 10
+
+	// Calling Start() after RegisterFilter() simulates a node restart after job creation, should reload Filter from db.
+	require.NoError(t, th.LogPoller.Start(th.Ctx))
+
+	latestBlock := finalityDepth + int64(numBlocks) + 1
+	require.NoError(t, th.LogPoller.Replay(th.Ctx, latestBlock))
+
+	// initializeLastProcessedBlock must return the unfulfilled VRF
+	// request block number (5) instead of finalizedBlockNumber (8)
+	lastProcessedBlock, err := th.Listener.initializeLastProcessedBlock(th.Ctx)
+	require.Nil(t, err)
+	require.Equal(t, int64(5), lastProcessedBlock)
+}
+
+func TestInitProcessedBlock_SomeUnfulfilledVRFReqs(t *testing.T) {
+	t.Parallel()
+
+	finalityDepth := int64(3)
+	th := setupVRFLogPollerListenerTH(t, false, finalityDepth, 3, 2, 1000, func(mockChain *evmmocks.Chain, curTH *vrfLogPollerListenerTH) {
+		mockChain.On("ID").Return(curTH.ChainID)
+		mockChain.On("LogPoller").Return(curTH.LogPoller)
+	})
+
+	// Block 3 to finalityDepth. Ensure we have finality number of blocks
+	for i := 1; i < int(finalityDepth); i++ {
+		th.Client.Commit()
+	}
+
+	// Emit some logs in blocks with VRF reqs interspersed
+	// No fulfillment for any VRF requests
+	numBlocks := 5
+	th.Client.Commit()
+	for i := 0; i < numBlocks; i++ {
+		_, err1 := th.Emitter.EmitLog1(th.Owner, []*big.Int{big.NewInt(int64(i))})
+		require.NoError(t, err1)
+		_, err1 = th.Emitter.EmitLog2(th.Owner, []*big.Int{big.NewInt(int64(i))})
+		require.NoError(t, err1)
+		th.Client.Commit()
+
+		// Create 2 blocks with VRF requests in each iteration
+		keyHash := [32]byte(th.Listener.job.VRFSpec.PublicKey.MustHash().Bytes())
+		preSeed := big.NewInt(105)
+		subID := uint64(1)
+		reqID1 := big.NewInt(int64(2 * i))
+		_, err2 := th.VRFLogEmitter.EmitRandomWordsRequested(th.Owner,
+			keyHash, reqID1, preSeed, subID, 10, 10000, 2, th.Owner.From)
+		require.NoError(t, err2)
+		th.Client.Commit()
+
+		reqID2 := big.NewInt(int64(2*i + 1))
+		_, err2 = th.VRFLogEmitter.EmitRandomWordsRequested(th.Owner,
+			keyHash, reqID2, preSeed, subID, 10, 10000, 2, th.Owner.From)
+		require.NoError(t, err2)
+		th.Client.Commit()
+	}
+
+	// Blocks till now: 2 (in SetupTH) + 2 (empty blocks) + 3*5 (EmitLog + VRF req/resp blocks) = 19
+
+	// Calling Start() after RegisterFilter() simulates a node restart after job creation, should reload Filter from db.
+	require.NoError(t, th.LogPoller.Start(th.Ctx))
+
+	latestBlock := finalityDepth + int64(numBlocks) + 1
+	require.NoError(t, th.LogPoller.Replay(th.Ctx, latestBlock))
+
+	// initializeLastProcessedBlock must return the earliest unfulfilled VRF request block
+	// number instead of finalizedBlockNumber
+	lastProcessedBlock, err := th.Listener.initializeLastProcessedBlock(th.Ctx)
+	require.Nil(t, err)
+	require.Equal(t, int64(7), lastProcessedBlock)
```

Comment:
shouldn't this be 6? block #5 has `Emitter.EmitLog1` and block #6 has the first RandomWordsRequested